spinalcordtoolbox/reports/qc.py

```diff
@@ -613,6 +612,17 @@ def update_description_file(self, dimension):
 
         :param: dimension 2-tuple, the dimension of the image frame (w, h)
         """
+        dest_path = self.qc_params.root_folder
+        html_path = os.path.join(dest_path, 'index.html')
+        # NB: We use 'r+' because it allows us to open an existing file for locking *without* immediately truncating the
+        # existing contents prior to opening. We can then use this file to overwrite the contents later in the function.
+        mode = "r+" if os.path.isfile(html_path) else "w"
```

Comment:
> My understanding is that a lock doesn't prevent opening and/or modifying the file; it only prevents another process from also locking the file. (That seems to be the case on Linux, but things could be very different on Windows.)

Your suspicions were correct. Opening a second file pointer in `'w'` mode causes `PermissionError: [Errno 13] Permission denied`. 😢 

However, I think there might be a simpler approach, specifically because of what you wrote here (emphasis mine): 

> Open the file with mode `'a'`, **which creates it if need be, doesn't erase it**, and allows us to grab a lock.

This made me realize that we're basically just looking for the `touch` command, but in Python, aren't we?

And, what do you know, `pathlib` has [`touch`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.touch):

> Create a file at this given path. [...] If the file already exists, the function succeeds if `exist_ok` is true (and its modification time is updated to the current time), otherwise [FileExistsError](https://docs.python.org/3/library/exceptions.html#FileExistsError) is raised.

Reply:
Actually, even simpler (since `touch` involves updating modification times, and maybe that ends up doing something weird, who knows).

```python
if not os.path.isfile(html_path):
    open(html_path, 'a').close()
```

Reply:
You don't even need the conditional:
```py3
# make sure the file exists
open(html_path, 'a').close()
```

Also, if you're opening in `'r+'` mode (which sounds like a good idea), make sure to call ['dest_file.truncate()`](https://docs.python.org/3/library/io.html#io.IOBase.truncate) after writing, otherwise there may be junk at the end:
```py3
dest_file = open(html_file, 'r+')
portalocker.lock(dest_file, portalocker.LOCK_EX)
...
dest_file.write(output)
dest_file.truncate()
dest_file.flush()
...
portalocker.unlock(dest_file)
dest_file.close()
```

Reply:
Aha, thank you! :)